Add internet recipe

sean-hut opened the pull request:
### Brief summary of what the package does

Provides a variety of ways to connect to the internet.

### Direct link to the package repository

https://github.com/sean-hut/internet

### Your association with the package

I am the package author and maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

sean-hut commented:
I searched for a package with this functionality and did not find one.

I search the following keywords with `M-x`, `C-h f`  and `apropos`.  I also searched MELPA.

Keywords:
- internet
- ethernet
- wireless
- wifi
- rfkill
- ip link
- wpa
- dhcp
- dhclient
- vpn

The only thing I found was `ovpn-mode` when I searched MELPA for the vpn keyword.  It has a different purpose than this package.

alphapapa commented:
Did you search for `network`?  I found:

    enwc is an available package.

         Status: Available from gnu -- Install
        Archive: gnu
        Version: 2.0
        Summary: The Emacs Network Client
       Requires: emacs-25.1
       Homepage: https://savannah.nongnu.org/p/enwc
       Keywords: external network wicd manager nm 

    ENWC is the Emacs Network Client.  It is designed to provide a front-end to
    various network managers, such as NetworkManager and Wicd.

alphapapa commented:
There's a lot of potentially dangerous code in this package, many calls to `sudo` this and `sudo` that using `async-shell-command` and non-shell-quoted arguments, like:
```
(defun internet-remove-wireless-configuration (configuration-file)
  "Remove a wpa_supplicant configuration file.
CONFIGURATION-FILE is the wpa_supplicant configuration file to remove."
  (interactive

   (list (completing-read
	  "wpa_supplicant configuration file to remove: "
	  (directory-files internet-wpa-supplicant-config-directory)
	  nil
	  t)))

  (let ((command (concat "sudo --stdin rm "
			 internet-wpa-supplicant-config-directory
			 configuration-file))

	(process-buffer "internet-remove-wireless-configuration"))

    (async-shell-command command process-buffer process-buffer)

    (save-excursion (delete-other-windows))))
```

sean-hut commented:
@alphapapa thank you for your response.

> Did you search for `network`?

No I overlooked that keyword.

I will also make sure to search the `gnu` archive in the future.

`enwc` serves a different purpose than this package.

`enwc` is a frontend for NetworkManager and Wicd.

This package is an alternative to NetworkManager's `nmcli` and Wicd's 
`wicd-cli`.

This package does not depend on either NetworkManager or Wicd.

This package can run in Emacs on the Linux console without a display
server like X or Wayland.

> There's a lot of potentially dangerous code in this package, many
> calls to `sudo` this and `sudo` that using `async-shell-command` and
> non-shell-quoted arguments

The three points to address are `sudo`, shell quoting and
`internet-remove-wireless-configuration`.

### `sudo`

The user gets to decide if they want to enter their sudo password when
they are prompted for it.

Here are the situations where `sudo` is used with
`async-shell-command` and the rationale for it.

1. Networking

Situation: When the commands `rfkill`, `ip link`, `wpa_supplicant` and
`dhclient` are run.

Rationale: The user may not have permission to run networking
commands.

Note: I just checked and `openvpn` looks like it can run without `sudo` so:

```
(defun internet--openvpn-connect-command (vpn)

  (concat "sudo --stdin openvpn --config "
	  (concat internet-openvpn-config-directory vpn ".ovpn")))
```

can be changed to:

```
(defun internet--openvpn-connect-command (vpn)

  (concat "openvpn --config "
	  (concat internet-openvpn-config-directory vpn ".ovpn")))
```

2. Killing a networking command

Situation: Killing `wpa_supplicant`, `dhclient` or `openvpn`.

Rationale: The user may not have permission to kill these processes.

3. `wpa_supplicant` configuration files

Situation: Adding and removing `wpa_supplicant` configuration files

Rationale: As the default directory for `wpa_supplicant` configuration
files is `/etc/wpa_supplicant/` the user may not have permission to
add or remove files from this directory.

### Shell Quoting

When you say

> non-shell-quoted arguments

are you referring to single and double quoting like this?

https://www.gnu.org/software/bash/manual/html_node/Quoting.html#Quoting

In this package there are two uses of unquoted shell arguments:
- configuration variables that have been manually set by the user
- arguments that the user entered when prompted

Could you please explain the benefit of using quotes in these cases.

This package has documentation covering the package configuration and
examples are provided.

https://sean-hut.github.io/internet/Configuration.html#Configuration

User enter values when prompt are constrained by:
- completions that require a match
- user configuration variables

The exception to constrained user arguments is the function
`internet-add-wireless-configuration`.

### `internet-remove-wireless-configuration`

Could you please clarify why you specifically point out
`internet-remove-wireless-configuration`.  I provide rationale for the
use of sudo in this case above in:
- `sudo` section
    - `wpa_supplicant` configuration files

riscy commented:
@alphapapa is referring to `shell-quote-arguments` which prevents people from doing something silly like setting `internet-wireless-interface` to ``` `rm important-file` ```, but it also escapes special characters like semicolons, pipes, etc., and you probably want to make these commands airtight if you're going to involve sudo.  By the way, in my quick glance at this code I notice you `defvar internet-wireless-interface nil` and then e.g.
```elisp
(defconst internet--ip-link-wireless-up-command
  (concat "sudo --stdin ip link set " internet-wireless-interface " up")
  "Command to set wireless interface to up using ip link.")
```
so even though `internet-wireless-interface` is a variable, it doesn't matter if the user changes it, the constant has already been set?

In the meantime, a quick pass over the checklist:

### internet.el ###

byte-compile (using Emacs 27.1):
```
internet.el:280:1:Warning: Unused lexical argument `process'
```
Just call it `_process` and the byte-compiler is happy.

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20200906.512):
```
7 issues found:
1:77: warning: You should depend on (emacs "24.1") if you need lexical-binding.
49:10: error: You should depend on (emacs "25.1") or the seq package if you need `seq'.
50:10: error: You should depend on (emacs "24.4") if you need `subr-x'.
288:5: error: You should depend on (emacs "24.4") if you need `string-trim-right'.
429:3: error: You should depend on (emacs "25.1") or the seq package if you need `seq-filter'.
430:16: error: You should depend on (emacs "24.4") if you need `string-suffix-p'.
434:18: error: You should depend on (emacs "24.4") if you need `string-remove-suffix'.
```

sean-hut commented:
@riscy thank you for your reply.

> shell-quote-arguments which prevents people from doing something
> silly like setting internet-wireless-interface to `rm
> important-file`, but it also escapes special characters like
> semicolons, pipes, etc., and you probably want to make these
> commands airtight if you're going to involve sudo.

Thanks for bringing `shell-quote-argument` to my attention.  In commit
2d5a814d1152ac56ed4fca7bc87eca7b97a220fb `shell-quote-argument` is
used to quote user provided arguments.

> you defvar internet-wireless-interface nil and then e.g.

```elisp
(defconst internet--ip-link-wireless-up-command
  (concat "sudo --stdin ip link set " internet-wireless-interface " up")
  "Command to set wireless interface to up using ip link.")
```

> so even though internet-wireless-interface is a variable, it doesn't
> matter if the user changes it, the constant has already been set?

I have changed the relevant defconsts to defvars in commit
a2ac78cec4b675d3363ca88cf3c59127e4c4ee86.

I am used to working with languages that enforce mutability
constraints.  One has to be extra careful when working with a language
like Elisp that does not have these helpful constraints.

Are there tools I can run on source code that would check for this
kind of mistake?

> byte-compile (using Emacs 27.1):

`internet.el:280:1:Warning: Unused lexical argument 'process'`

> Just call it _process and the byte-compiler is happy.

Thanks for sharing how the underscore can help with an unused lexical
argument.  I was not sure how to deal with that warning.  This warning
is addressed in commit 13e86eee638f843a53f6695ffdf4d9203906041e.

> package-lint (using version 20200906.512):

```
7 issues found:
1:77: warning: You should depend on (emacs "24.1") if you need lexical-binding.
49:10: error: You should depend on (emacs "25.1") or the seq package if you need `seq'.
50:10: error: You should depend on (emacs "24.4") if you need `subr-x'.
288:5: error: You should depend on (emacs "24.4") if you need `string-trim-right'.
429:3: error: You should depend on (emacs "25.1") or the seq package if you need `seq-filter'.
430:16: error: You should depend on (emacs "24.4") if you need `string-suffix-p'.
434:18: error: You should depend on (emacs "24.4") if you need `string-remove-suffix'.
```

In commit dad27f147d047f83ff44f50a75e803b79ec8c5f7 I added Emacs 25.1
as a package requirement to address those `package-lint` errors.

I did not see these `package-lint` errors because I was stuck on an
error about a `dash` requirement.  This seemed like a strange error to
me because I had refactored this package so that it no longer required
`dash`.

I am running `package-lint` through `elisp-lint` to avoid it's
incompatibility with the `straight` package manager.  The requirement
for `dash` was a requirement of `elisp-lint`.  I solved that error and
then I could see these dependency errors.

I will try to make sure that I have resolved all the linting error and
warning before I make future package submissions.

riscy commented:
> Are there tools I can run on source code that would check for this kind of mistake?

Not that I'm aware of.  Correct me if I'm wrong but I think a similar issue persists with these three definitions happening in sequence:
```elisp
(defvar internet-rfkill-device-id nil
  "Device ID for rfkill.")

(defvar internet--rfkill-unblock-command
  (concat "sudo --stdin rfkill unblock "
	  (shell-quote-argument internet-rfkill-device-id))
  "Command to soft unblock wireless device using rfkill.")

(defvar internet--rfkill-block-command
  (concat "sudo --stdin rfkill block "
	  (shell-quote-argument internet-rfkill-device-id))
  "Command to soft block wireless device using rfkill.")
```
If a user modifies internet-rfkill-device-id, the downstream variables won't be updated to reflect the new value.  I think what I'd recommend instead is to use a `defun` for `internet--rfkill-(un)block-command` so that it's always reading the freshest values.  Also note that the above code will actually not run on a fresh Emacs instance because `(shell-quote-argument nil)` throws an error.

It's a future ambition of MELPA to help do some of the linting for users.

riscy commented:
I'll also add that I'm not able to make sense of the `internet-rfkill-device-id` variable -- it isn't modified anywhere and isn't mentioned in the Commentary block.

riscy commented:
Friendly ping. :)

sean-hut commented:
@riscy

> Friendly ping. :)

Thanks

---

> I think a similar issue persists with these three definitions
> happening in sequence:

```elisp
(defvar internet-rfkill-device-id nil
  "Device ID for rfkill.")

(defvar internet--rfkill-unblock-command
  (concat "sudo --stdin rfkill unblock "
	  (shell-quote-argument internet-rfkill-device-id))
  "Command to soft unblock wireless device using rfkill.")

(defvar internet--rfkill-block-command
  (concat "sudo --stdin rfkill block "
	  (shell-quote-argument internet-rfkill-device-id))
  "Command to soft block wireless device using rfkill.")
```

> If a user modifies internet-rfkill-device-id, the downstream
> variables won't be updated to reflect the new value. I think what
> I'd recommend instead is to use a defun for 
> internet--rfkill-(un)block-command so that it's always reading the
> freshest values.

Thank you for the suggestion.  I have changed the relevant `defvar`s
to `defun`s throughout this package.

Commit: 83a553647e13da44ef3311279cfaeb48e5def3df

---

> Also note that the above code will actually not run on a fresh Emacs
> instance because (shell-quote-argument nil) throws an error.

I have changed the variables that get shell quoted to default string
values.  Users will have to configure these variables for this package
to work for them.

Commit: f18f38e009c4f1d7d6194b5e64b60dfa683b5b0b

---

> I'll also add that I'm not able to make sense of the
> internet-rfkill-device-id variable -- it isn't modified anywhere and
> isn't mentioned in the Commentary block.

The internet-rfkill-device-id variable is explained in the
documentation for the package.
https://sean-hut.github.io/internet/Configuration.html#Configure-for-wireless-connecting

---

Is the `awaiting-upstream` label for the FSF license review?

riscy commented:
Removed it now - I put that label there originally while waiting to hear back, not (necessarily) because of the license.

riscy commented:
So, yes, as we talked about [here](https://github.com/melpa/melpa/pull/7189#issuecomment-735508725), this license is not on the GPL-Compatible list.  I can't speak for other MELPA maintainers, but I'm not equipped to get into the legal nuances of why a license is or isn't on that list even if the difference is down to a couple words.  But we _do_ want compatibility with the GPL for packages in MELPA.

I don't mean to sound harsh, but please do _not_ check "the package is released under a GPL-Compatible Free Software License." unless/until your license is actually on that list.

I know from your other PR that you are firm on wanting that license, which is fine, but I'll mark this awaiting upstream and not look any more closely until that's the case.

riscy commented:
Per @tarsius this license is okay -- also I apologize for the long turnaround time.

Overall I think the use of shell-quote-argument looks good now.  It can't hurt to be too careful when sudo is involved.  Looking at it with fresh eyes, I see a couple more issues that could be patched:

- The use of `(save-excursion (delete-other-windows))` in several places will disrupt user's window configurations if they'd already split the pane; and why not leave the window intact, so users can watch for any warnings or errors?
- Strings don't need to be quoted, e.g. `'"sudo --stdin pkill wpa_supplicant"`
- This is not strictly a blocker, but there is lots of room for deduplication.  Decreasing the code surface area of a package that makes shell commands through `sudo` is a good idea.  For example if I put `internet-wireless-connect` and `internet-wireless-openvpn-connect` next to each other I see they're virtually identical.  And if you deduplicated them, the helper functions like `internet--rfkill-unblock-command` and ``internet--ip-link-wireless-up-command` would only have one caller and could be inlined, making the sequence of commands much more transparent.

riscy commented:
I'll mark this awaiting upstream until we hear back.

riscy commented:
Friendly ping. :)

I'm realizing some of my feedback above may have been a bit imprecise and hard to act on, but let me know what you think -- I think the first one is the most important to consider.
